### PR TITLE
feat: implement lsp features for effects

### PR DESCRIPTION
**Pull request by mlutze**

### CODE REVIEW

**Comment by magnus-madsen** [comment]
We have to find a name that is not `Ref`. That's too confusing!

**Comment by mlutze** [comment]
> We have to find a name that is not `Ref`. That's too confusing!

Any suggestions? Maybe:
* Box
* Use
* Mention

**Comment by magnus-madsen** [comment]
> > We have to find a name that is not `Ref`. That's too confusing!
> 
> Any suggestions? Maybe:
> 
>     * Box
> 
>     * Use
> 
>     * Mention

Hmm.

**Comment by magnus-madsen** [comment]
Bring out the ThesauruSAUR!

image

**Comment by mlutze** [comment]
Screenshot from 2022-07-11 09-35-21

File: main/src/ca/uwaterloo/flix/language/ast/Ast.scala [hunk 1]
```diff
@@ -495,6 +495,20 @@ object Ast {
...
+    case class Ref(sym: Symbol.EffectSym, loc: SourceLocation)
+  }
+
+  case object Op {
+    /**
+      * Represents a reference to an effect operation.
+      */
+    case class Ref(sym: Symbol.OpSym, loc: SourceLocation)
+  }
+
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/Ast.scala:511]: What do you think about simply `EffectSymUse` and `OpSymUse`? (not nested in case objects)

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/Ast.scala:511]: Lets do that then.

File: main/src/ca/uwaterloo/flix/language/phase/Resolver.scala [hunk 2]
```diff
@@ -969,33 +969,40 @@ object Resolver {
           val eVal = visitExp(exp, region)
           val fVal = lookupEffect(eff, ns0, root)
           mapN(eVal, fVal) {
-            case (e, f) => ResolvedAst.Expression.Without(e, f.sym, loc)
+            case (e, f) =>
+              val effRef = Ast.EffectSymUse(f.sym, eff.loc)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Resolver.scala:973]: local variable name

File: main/src/ca/uwaterloo/flix/language/phase/Resolver.scala [hunk 3]
```diff
@@ -969,33 +969,40 @@ object Resolver {
...
               val rulesVal = traverse(rules) {
                 case NamedAst.HandlerRule(ident, fparams, body) =>
                   val opVal = findOpInEffect(ident, f)
                   val fparamsVal = resolveFormalParams(fparams, taenv, ns0, root)
                   val bodyVal = visitExp(body, region)
                   mapN(opVal, fparamsVal, bodyVal) {
-                    case (o, fp, b) => ResolvedAst.HandlerRule(o.sym, fp, b)
+                    case (o, fp, b) =>
+                      val opRef = Ast.OpSymUse(o.sym, ident.loc)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Resolver.scala:990]: local variable name

**Comment by magnus-madsen** [comment]
Thanks!

**Pull request merged**